core/commons/che-core-commons-observability/src/main/java/org/eclipse/che/commons/observability/MeteredExecutorServiceWrapper.java

```diff
@@ -0,0 +1,102 @@
+ */
+@Singleton
+public class MeteredExecutorServiceWrapper implements ExecutorServiceWrapper {
+  private final MeterRegistry meterRegistry;
+
+  @Inject
+  public MeteredExecutorServiceWrapper(MeterRegistry meterRegistry) {
+    this.meterRegistry = meterRegistry;
+  }
+
+  @Override
+  public ScheduledExecutorService wrap(
+      ScheduledExecutorService executor, String name, String... tags) {
+
+    monitorThreadPoolExecutor(executor, name, tags);
+    return ExecutorServiceMetrics.monitor(meterRegistry, executor, name, Tags.of(tags));
+  }
+
+  @Override
+  public ExecutorService wrap(ExecutorService executor, String name, String... tags) {
+    monitorThreadPoolExecutor(executor, name, tags);
+    return ExecutorServiceMetrics.monitor(meterRegistry, executor, name, Tags.of(tags));
+  }
+
+  @Override
+  public CronExecutorService wrap(CronExecutorService executor, String name, String... tags) {
+    monitorThreadPoolExecutor(executor, name, tags);
+    new io.micrometer.core.instrument.binder.jvm.ExecutorServiceMetrics(
+            executor, name, Tags.of(tags))
+        .bindTo(meterRegistry);
+
+    return new TimedCronExecutorService(meterRegistry, executor, name, Tags.of(tags));
+  }
+
+  private void monitorThreadPoolExecutor(ExecutorService executor, String name, String... tags) {
+    String className = executor.getClass().getName();
+    ThreadPoolExecutor unwrappedThreadPoolExecutor = null;
+    if (executor instanceof ThreadPoolExecutor) {
+      unwrappedThreadPoolExecutor = (ThreadPoolExecutor) executor;
+    } else if (className.equals(
+        "java.util.concurrent.Executors$DelegatedScheduledExecutorService")) {
+      unwrappedThreadPoolExecutor = unwrapThreadPoolExecutor(executor, executor.getClass());
+    } else if (className.equals(
+        "java.util.concurrent.Executors$FinalizableDelegatedExecutorService")) {
+      unwrappedThreadPoolExecutor =
+          unwrapThreadPoolExecutor(executor, executor.getClass().getSuperclass());
+    }
+    if (unwrappedThreadPoolExecutor != null) {
+      CountedThreadFactory.monitorThreads(
+          meterRegistry, unwrappedThreadPoolExecutor, name, Tags.of(tags));
+      CountedRejectedExecutionHandler.monitorRejections(
+          meterRegistry, unwrappedThreadPoolExecutor, name, Tags.of(tags));
+    }
+  }
+  /**
+   * Every ScheduledThreadPoolExecutor created by {@link Executors} is wrapped. Also, {@link
+   * Executors#newSingleThreadExecutor()} wrap a regular {@link ThreadPoolExecutor}.
+   */
+  @Nullable
+  private ThreadPoolExecutor unwrapThreadPoolExecutor(ExecutorService executor, Class<?> wrapper) {
+    try {
+      Field e = wrapper.getDeclaredField("e");
+      e.setAccessible(true);
+      return (ThreadPoolExecutor) e.get(executor);
+    } catch (NoSuchFieldException | IllegalAccessException e) {
+      // Do nothing. We simply can't get to the underlying ThreadPoolExecutor.
```

Comment:
maybe log a warning?

Reply:
If you insist I think we can add this. However, I would prefer to keep the original code (taken from ExecutorServiceMetrics) intact. Moreover, it should go away when we will reuse micrometer  with this pr https://github.com/micrometer-metrics/micrometer/pull/1671

Reply:
What happens when we hit this `catch`? You've chose fail fast in `TimedCronExecutorService.java` and this looks like silent failure. I'm wondering why these 2 different ways here.
Overall, I'd like to see more logging throughout whole codebase, it would be useful to investigate crashes. Silently eating exceptions without any clue might be dangerous.

Reply:
Added more logging.